### ISSUE TITLE
 Unbound variable: ConceptNode Afer running (use-modules (opencog)) in SchemeEval

### ISSUE THREAD

**Issue by AmeBel** [opencog/atomspace#508]
What am I missing?

#### Test class
```c++
class ActionUTest : public CxxTest::TestSuite
{
private:
    AtomSpace _as;
    SchemeEval _scm;
public:
    ActionUTest() : _scm(&_as)
    {
        // Print output to standard
        logger().setLevel(Logger::DEBUG);
        logger().setPrintToStdoutFlag(true);
        logger().setPrintLevelFlag(true);
    }

    ~ActionUTest()
    {
        // erase the log file if no assertions failed
        //if (!CxxTest::TestTracker::tracker().suiteFailed())
        //    std::remove(logger().getFilename().c_str());
    }

    void setUp()
    {
        cout <<_scm.eval("(add-to-load-path \"/usr/local/share/opencog/scm\")");
        cout << _scm.eval("(use-modules (opencog) (opencog rule-engine))");
        cout << _scm.eval("(load \"test-rbs.scm\")");
    }


    // Why this check? B/c PatternLink api's behavior might change.
    // What is being checked? That the derived rule doesn't have any form of
    // virtual-terms.
    void test_constuctor() {
   ...
   ...
    }
};
```

#### Content of `test-rbs.scm`
```
(ConceptNode "abc")
```

#### Output of running the test
```bash
opencog@4fd61e755acb:/opencog/build/tests/planning$ ctest -VV ActionUTest
UpdateCTestConfiguration  from :/opencog/build/tests/planning/DartConfiguration.tcl
UpdateCTestConfiguration  from :/opencog/build/tests/planning/DartConfiguration.tcl
Test project /opencog/build/tests/planning
Constructing a list of tests
Done constructing a list of tests
Checking test dependency graph...
Checking test dependency graph end
test 1
    Start 1: ActionUTest

1: Test command: /opencog/build/tests/planning/ActionUTest
1: Test timeout computed to be: 9.99988e+06
1: 
1: 
1: Backtrace:
1: In ice-9/boot-9.scm:
1:  157: 13 [catch #t #<catch-closure 24f5c60> ...]
1: In unknown file:
1:    ?: 12 [apply-smob/1 #<catch-closure 24f5c60>]
1: In ice-9/boot-9.scm:
1:  157: 11 [catch #t #<catch-closure 24f5b40> ...]
1: In unknown file:
1:    ?: 10 [apply-smob/1 #<catch-closure 24f5b40>]
1:    ?: 9 [call-with-input-string "(load \"test-rbs.scm\")" ...]
1: In ice-9/boot-9.scm:
1: 2320: 8 [save-module-excursion #<procedure 255f8d0 at ice-9/eval-string.scm:65:9 ()>]
1: In ice-9/eval-string.scm:
1:   44: 7 [read-and-eval #<input: string 2428680> #:lang ...]
1:   37: 6 [lp (load "test-rbs.scm")]
1: In ice-9/boot-9.scm:
1: 2320: 5 [save-module-excursion #<procedure 2563e40 at ice-9/boot-9.scm:3961:3 ()>]
1: 3968: 4 [#<procedure 2563e40 at ice-9/boot-9.scm:3961:3 ()>]
1: In unknown file:
1:    ?: 3 [load-compiled/vm "/home/opencog/.cache/guile/ccache/2.0-LE-8-2.0/opencog/build/tests/planning/test-rbs.scm.go"]
1: In /opencog/build/tests/planning/test-rbs.scm:
1:    1: 2 [#<procedure 24f5160 at /opencog/build/tests/planning/test-rbs.scm:1:0 ()>]
1: In ice-9/boot-9.scm:
1:  102: 1 [#<procedure 2563f40 at ice-9/boot-9.scm:97:6 (thrown-k . args)> unbound-variable ...]
1: In unknown file:
1:    ?: 0 [apply-smob/1 #<catch-closure 24f5b00> unbound-variable ...]
1: 
1: ERROR: In procedure apply-smob/1:
1: ERROR: In procedure module-lookup: Unbound variable: ConceptNode
1: ABORT: unbound-variable
1: Running cxxtest tests (1 test).OK!
1/1 Test #1: ActionUTest ......................   Passed    0.12 sec

100% tests passed, 0 tests failed out of 1

Total Test time (real) =   0.13 sec

```

**Comment by misgeatgit** [comment]
Perhaps you should load the core_types.scm file so that guile knows what a ConceptNode is.

**Comment by AmeBel** [comment]
That has no effect.  Plus, `(use-modules (opencog) ...` should suffice, right?

**Comment by misgeatgit** [comment]
Oh, I just wanted to suggest loading utilities.scm not core_types.scm as you have tried it. Does that make difference? Otherwise am also confused why that error is thrown.

**Comment by AmeBel** [comment]
Those scm files are included in the `opencog` module. And the error isn't related to the functions in utilities.scm . But, in case it is, with some freaks of c++, I tried it and has no difference.

By the way, I reported this here assuming it is an issue with SchemeEval, but the actual class is being developed in the opencog/opencog repo.

**Comment by linas** [comment]
Try putting `(use-modules (opencog) )` into `test-rbs.scm` directly. That should fix it.

**Comment by linas** [comment]
I've seen similar weird bugs e.g. `stv` not found, and I don't understand them.  Its as if stuff between two different calls of `_scm.eval()` is forgotten, or a later definition hides an earlier one.  May have something to do with the "evaluation environment".

**Comment by linas** [comment]
this is the cleanest reproduce of the bug ever, it might actually allow it to be debugged ..

**Comment by linas** [comment]
debugging now. I used to get this kind of message when the module itself had an error in it -- it was silently failing to load, and so the things it defined were undefined.

**Comment by linas** [comment]
think I found it.

**Comment by linas** [comment]
Ugh. This was very confusing.  First: the "fix" add this to your test case:
```
std::cout << _scm.eval("(use-modules (opencog))");
std::cout << _scm.eval("(load-from-path \"opencog.scm\")");
```

Here's why it fixes it. When you create the SchemeEval C++ class, it causes lots of things to happen -- initialization of scheme, declaration of lots of opencog internals.  One of the things that runs is this:
```
SchemeSmob.cc:	scm_c_define_module("opencog", register_procs, NULL);
SchemeSmob.cc:	scm_c_use_module("opencog");
```

To guile, the second statement `scm_c_use_module("opencog");` is the "same thing" as saying `(use-modules (opencog))`  -- except that it did actually not load `opencog.scm`, instead it just ran the C++-related init.  Then later on, when you say `(use-modules (opencog))`  guile thinks to itself "oh gee, I already loaded that, I won't do it again", and thus, nothing at all in the  `opencog.scm` file is run.  

That's why you got the error, and that's why the work-around works.  I'm not sure what the correct long-term fix is .. maybe we should not call scm_c_define ... testing now.

**Comment by linas** [comment]
I give up. Nothing to be done. If you want to use guile from c++, you will have to either use the ugly SCM_PRELOAD mechanism, or say `_scm.eval("(load-from-path \"opencog.scm\")");`  -- the problem with the latter is that it doesn't play nice with unit tests, where `opencog.scm` might not be installed yet, so you have to make sure the build and source directories are in the load-path. Yuck.  Oh, and they can't be relative, because then the unit test won't run in any old place. Double-yuck.  You would then have to add 3 or 4 different `add-to-load-path` statements to make sure all sorts of different paths get searched.  I find it all very annoying.  Still, maybe its still better than the SCM_PRELOAD thing...

**Comment by AmeBel** [comment]
@linas Here is a propsal and a partial implementation for a solution, https://github.com/opencog/atomspace/pull/511.

**Comment by linas** [comment]
See also bug opencog/opencog#2021 -- I am starting to think that calling `scm_c_define_module()` from C++ code is a mistake ... in that bug, it seems to interfere with the discovery and loading of the R2L rules.  Of course, the root cause of that bug is not known, so maybe I am jumping to conclusions.

**Comment by linas** [comment]
The "temporary hack" in `gen-r2l-en-rulebase.scm` is completely breaking lots of code: it basicly ends up defininig the entire universe in the rule-engine module, and thus, almost nothing is defined outside of the rule-engine module.  I am trying to fix this now.

**Comment by linas** [comment]
I think that this pull-req fixes this problem:  #703  (It doesn't merge right now due to jenkins, can't tell what's wrong).

**Comment by linas** [comment]
closing, I'm guessing this issue is fixed.
